test/utils/Checkpoints.t.sol

```diff
@@ -0,0 +1,276 @@
+
+        // initial state
+        assertTrue(_ckpts.length() == 0);
+        assertTrue(_ckpts.latest() == 0);
+        _assertLatestCheckpoint(false, 0, 0);
+
+        uint256 duplicates = 0;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint32 key = keys[i];
+            uint224 value = values[i % values.length];
+            if (i > 0 && key == keys[i - 1]) ++duplicates;
+
+            // push
+            _ckpts.push(key, value);
+
+            // check length & latest
+            assertTrue(_ckpts.length() == i + 1 - duplicates);
+            assertTrue(_ckpts.latest() == value);
+            _assertLatestCheckpoint(true, key, value);
+        }
+    }
+
+    function testLookup(uint32[] memory keys, uint224[] memory values, uint32 lookup) public {
+        vm.assume(values.length > 0);
+        _prepareKeys(keys, 64);
+
+        uint32 lastKey = keys.length == 0 ? 0 : keys[keys.length - 1];
+        lookup = _boundUint32(lookup, 0, lastKey + 64);
+
+        uint224 upper = 0;
+        uint224 lower = 0;
+        uint32 lowerKey = type(uint32).max;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint32 key = keys[i];
+            uint224 value = values[i % values.length];
+
+            // push
+            _ckpts.push(key, value);
+
+            // track expected result of lookups
+            if (key <= lookup) {
+                upper = value;
+            }
+            if (key >= lookup && (i == 0 || keys[i - 1] < lookup)) {
+                lowerKey = key;
+            }
+            if (key == lowerKey) {
+                lower = value;
+            }
+        }
+
+        // check lookup
+        assertTrue(_ckpts.lowerLookup(lookup) == lower);
+        assertTrue(_ckpts.upperLookup(lookup) == upper);
+        assertTrue(_ckpts.upperLookupRecent(lookup) == upper);
+    }
+}
+
+contract CheckpointsTrace160Test is Test {
+    using Checkpoints for Checkpoints.Trace160;
+
+    Checkpoints.Trace160 internal _ckpts;
+
+    // helpers
+    function _boundUint96(uint96 x, uint96 min, uint96 max) internal view returns (uint96) {
+        return SafeCast.toUint96(bound(uint256(x), uint256(min), uint256(max)));
+    }
+
+    function _prepareKeys(uint96[] memory keys, uint96 maxSpread) internal view {
+        uint96 lastKey = 0;
+        for (uint256 i = 0; i < keys.length; ++i) {
+            uint96 key = _boundUint96(keys[i], lastKey, lastKey + maxSpread);
+            keys[i] = key;
+            lastKey = key;
+        }
+    }
+
+    function _assertLatestCheckpoint(bool exist, uint96 key, uint160 value) internal {
+        (bool _exist, uint96 _key, uint160 _value) = _ckpts.latestCheckpoint();
+        assertTrue(_exist == exist);
+        assertTrue(_key == key);
+        assertTrue(_value == value);
+    }
+
+    // tests
+    function testPush(uint96[] memory keys, uint160[] memory values) public {
+        vm.assume(values.length > 0);
+        _prepareKeys(keys, 64);
```

Comment:
Why 64 in both tests? I see it makes sense in the lookup tests because of the binary search but I don't see a good reason for using it here.

Reply:
There is no real point to having crazy gaps between checkpoints, and I we need the keys to not overflow uint32. I also want to get some key duplication, which is not likely to happen if the allowed space is too big.

Reply:
64 seems fine to me. But I would define it as a constant at the top of the file and explain the rationale.

Reply:
> @ernestognw The Solidity generation scripts run Prettier on the generated code.
> 
> https://github.com/OpenZeppelin/openzeppelin-contracts/blob/cf86fd9962701396457e50ab0d6cc78aa29a5ebc/scripts/generate/run.js#L36
> 
> Does this address your concern?

Yes it does, thanks!

Reply:
Agree, let's make it a constant with a comment.